### PR TITLE
Require `@Incremental` for incremental inputs

### PR DESCRIPTION
**Pull request by wolfs**
Only properties annotated with `@Incremental` can be queried for
incremental changes. The other properties are considered
non-incremental and changes to those cause the unit of work run
non-incrementally.

### CODE REVIEW

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/project/taskfactory/TaskPropertyNamingIntegrationTest.groovy [hunk 1]
```diff
@@ -81,7 +81,7 @@ class TaskPropertyNamingIntegrationTest extends AbstractIntegrationSpec {
                     def inputFiles = [:]
                     TaskPropertyUtils.visitProperties(project.services.get(PropertyWalker), it, new PropertyVisitor.Adapter() {
                         @Override
-                        void visitInputFileProperty(String propertyName, boolean optional, boolean skipWhenEmpty, Class<? extends FileNormalizer> fileNormalizer, PropertyValue value, InputFilePropertyType filePropertyType) {
+                        void visitInputFileProperty(String propertyName, boolean optional, boolean skipWhenEmpty, boolean isIncrementalInput, Class<? extends FileNormalizer> fileNormalizer, PropertyValue value, InputFilePropertyType filePropertyType) {
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/project/taskfactory/TaskPropertyNamingIntegrationTest.groovy]: Nit: `incremental` should be okay.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy [hunk 2]
```diff
@@ -61,6 +77,22 @@ class IncrementalInputsIntegrationTest extends AbstractIncrementalTasksIntegrati
         return ChangeTypeInternal.ADDED
     }
 
+    @Override
+    String getPrimaryInputAnnotation() {
+        return "@${Incremental.simpleName}"
+    }
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy]: Why is this called the primary input annotation?

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy [hunk 3]
```diff
@@ -73,4 +105,45 @@ class IncrementalInputsIntegrationTest extends AbstractIncrementalTasksIntegrati
...
 
+    def "cannot query non-incremental file input parameters"() {
+        given:
+        previousExecution()
+
+        when:
+        file("inputs/new-input-file.txt") << "new file"
+        buildFile << """
+            tasks.withType(IncrementalTask).configureEach {
+                queryChangesFor.add("nonIncrementalInput")
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy]: This indirection is a bit hard to follow. Can we maybe pass actual code to execute into the task to do the querying instead?

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy [hunk 4]
```diff
@@ -73,4 +105,45 @@ class IncrementalInputsIntegrationTest extends AbstractIncrementalTasksIntegrati
...
+    }
+
+    def "changes to non-incremental input parameters cause a rebuild"() {
+        given:
+        file("nonIncrementalInput").makeOlder()
+        previousExecution()
+
+        when:
+        file("inputs/new-input-file.txt") << "new file"
+        file("nonIncrementalInput").text = 'changed'
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy]: It would be good to split this into two tests to see each change triggering non-incremental execution.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy [hunk 5]
```diff
@@ -73,4 +105,45 @@ class IncrementalInputsIntegrationTest extends AbstractIncrementalTasksIntegrati
...
+        then:        
+        executesWithRebuildContext("ext.added += ['new-input-file.txt']")
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy]: This took a bit for me to decipher. For one, `ext.added` doesn't really convey that these are the expectations we set. This would be much more readable for me:

```groovy
executesNonIncrementally() // when there are no additional files to expect
executesNonIncrementallyWithAdditionalInputs('new-input-file.txt') // when there are no additional files to expect
```

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy [hunk 6]
```diff
@@ -73,4 +105,45 @@ class IncrementalInputsIntegrationTest extends AbstractIncrementalTasksIntegrati
...
+        setupTaskSources("@${SkipWhenEmpty.simpleName}")
+
+        given:
+        previousExecution()
+
+        when:
+        file('inputs/file1.txt') << "changed content"
+
+        then:
+        executesWithIncrementalContext("ext.modified = ['file1.txt']")
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy]: I'd also refactor this to:

```groovy
executesIncrementally(modified: ['file1.txt'])
```

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java [hunk 7]
```diff
@@ -289,11 +291,14 @@ public void outputsRemovedAfterFailureToLoadFromCache() {
         }
 
         @Override
-        public void visitFileInputs(InputFilePropertyVisitor visitor) {
-            for (InputFilePropertySpec inputFileProperty : context.getTaskProperties().getInputFileProperties()) {
+        public void visitInputFileProperties(InputFilePropertyVisitor visitor) {
+            boolean requiresInputChanges = isRequiresInputChanges();
+            boolean usesIncrementalTaskInputs = requiresInputChanges && isUsesIncrementalTaskInputs();
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: I would mention in the names here that this is about legacy. I would also mention the effect of using `IncrementalTaskInputs` being that everything is considered incremental.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java [hunk 8]
```diff
@@ -289,11 +291,14 @@ public void outputsRemovedAfterFailureToLoadFromCache() {
...
+            ImmutableSortedSet<InputFilePropertySpec> inputFileProperties = context.getTaskProperties().getInputFileProperties();
+            for (InputFilePropertySpec inputFileProperty : inputFileProperties) {
                 Object value = inputFileProperty.getValue();
                 if (value != null) {
-                    visitor.visitInputFileProperty(inputFileProperty.getPropertyName(), value);
+                    visitor.visitInputFileProperty(inputFileProperty.getPropertyName(), value, usesIncrementalTaskInputs || inputFileProperty.isIncremental() || inputFileProperty.isSkipWhenEmpty());
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: That condition definitely deserves its own line being assigned to a named variable, probably even its own static function.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/AbstractFingerprintChanges.java [hunk 9]
```diff
@@ -16,18 +16,19 @@
...
 import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
 import org.gradle.internal.fingerprint.FileCollectionFingerprint;
 
+import java.util.SortedMap;
+
 public abstract class AbstractFingerprintChanges implements ChangeContainer {
-    protected final ImmutableSortedMap<String, FileCollectionFingerprint> previous;
-    protected final ImmutableSortedMap<String, CurrentFileCollectionFingerprint> current;
+    protected final SortedMap<String, FileCollectionFingerprint> previous;
+    protected final SortedMap<String, CurrentFileCollectionFingerprint> current;
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/AbstractFingerprintChanges.java]: Why change this?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java [hunk 10]
```diff
@@ -147,8 +149,8 @@ public IncrementalDetectedExecutionStateChanges(
         }
 
         @Override
-        public InputChangesInternal createInputChanges(ImmutableMultimap<Object, String> incrementalParameterNameByValue) {
-            return new IncrementalInputChanges(inputFileChanges, incrementalParameterNameByValue);
+        public InputChangesInternal createInputChanges(ImmutableMultimap<Object, String> incrementalParameterNamesByValue) {
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java]: Does this still need to be a multimap? We said we'd fail on collision earlier.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java [hunk 11]
```diff
@@ -61,10 +62,11 @@ public ExecutionStateChanges detectChanges(AfterPreviousExecutionState lastExecu
             thisExecution.getInputFileProperties(),
             "Input file",
             executable);
-        InputFileChanges directInputFileChanges = new DefaultInputFileChanges(
+        DefaultInputFileChanges allInputFileChanges = new DefaultInputFileChanges(
             lastExecution.getInputFileProperties(),
             thisExecution.getInputFileProperties());
-        InputFileChanges inputFileChanges = errorHandling(executable, caching(directInputFileChanges));
+        InputFileChanges nonIncrementalInputFileChanges = allInputFileChanges.nonIncrementalChanges(incrementalPropertyNames);
+        InputFileChanges incrementalInputFileChanges = errorHandling(executable, caching(allInputFileChanges.incrementalChanges(incrementalPropertyNames)));
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java]: Why no error handling for the non-incremental changes?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java [hunk 12]
```diff
@@ -61,10 +62,11 @@ public ExecutionStateChanges detectChanges(AfterPreviousExecutionState lastExecu
             thisExecution.getInputFileProperties(),
             "Input file",
             executable);
-        InputFileChanges directInputFileChanges = new DefaultInputFileChanges(
+        DefaultInputFileChanges allInputFileChanges = new DefaultInputFileChanges(
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java]: This smells weird: we create an object of type `DefaultInputFileChanges`, but we never use it as `InputFileChanges`. Instead, it has two methods that return (potentially new) instances of `DefaultInputFileChanges`. Maybe we should inline what happens in here?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultInputFileChanges.java [hunk 13]
```diff
@@ -16,15 +16,19 @@
...
 import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
 import org.gradle.internal.fingerprint.FileCollectionFingerprint;
 
+import java.util.SortedMap;
+
 public class DefaultInputFileChanges extends AbstractFingerprintChanges implements InputFileChanges {
     private static final String TITLE = "Input";
 
-    public DefaultInputFileChanges(ImmutableSortedMap<String, FileCollectionFingerprint> previous, ImmutableSortedMap<String, CurrentFileCollectionFingerprint> current) {
+    public DefaultInputFileChanges(SortedMap<String, FileCollectionFingerprint> previous, SortedMap<String, CurrentFileCollectionFingerprint> current) {
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultInputFileChanges.java]: Why this change again?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/InputFileChanges.java [hunk 14]
```diff
@@ -16,9 +16,23 @@
...
+    InputFileChanges EMPTY = new InputFileChanges() {
+
+        @Override
+        public boolean accept(ChangeVisitor visitor) {
+            return true;
+        }
+
+        @Override
+        public boolean accept(String propertyName, ChangeVisitor visitor) {
+            throw new InvalidUserDataException("Cannot query incremental changes: No incremental properties declared.");
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/InputFileChanges.java]: We should probably report the property name here.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy]: The annotation is placed on the primary input of the task, the input directory. Do you have a better name for this input property?

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy (hunk 4 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy]: I don't understand. Only the second change causes a rebuild, the incremental change doesn't. Note that we have some more tests in the base class, which check changes on the incremental input already.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/AbstractFingerprintChanges.java (hunk 9 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/AbstractFingerprintChanges.java]: We create two kinds of filtered changes from the inputs: the incremental one and the non-incremental one. For the non-incremental one, we only need to iterate once and don't need to query, so I figured we can use `Maps.filterKeys` to do the (lazy) filtering on the backing maps and save some cycles when we don't need it.

`Maps.filterKeys` only returns a `SortedMap`, which is sufficient for comparing the changes.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java (hunk 10 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java]: I'll change it.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java (hunk 11 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java]: The non-incremental changing will be added to an `errorHandling`, `summarizing` changes later down. That's why I don't need the error handling here.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java (hunk 12 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/DefaultExecutionStateChangeDetector.java]: Good point! I moved both methods to the change detector.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java (hunk 7 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: Done.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java (hunk 8 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: Done.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/InputFileChanges.java (hunk 14 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/InputFileChanges.java]: Done.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalInputsIntegrationTest.groovy]: I created a new task instead of trying to cram everything in the one task.
